[Dart 3] Patterns-related control flow (#4780)
dart-lang/site-www: src/language/branches.md

---
title: Branches 
description: Learn how to use branches to control the flow of your Dart code.
---

This page shows how you can control the flow of your Dart code using branches:

- `if` statements and elements
- `if-case` statements and elements
- `switch` statements and expressions

You can also manipulate control flow in Dart using [loops][], like `for`
and `while`, or even using [exceptions][], like `try-catch` and `throw`.

## If

Dart supports `if` statements with optional `else` clauses. The condition in
parentheses after `if` must be an expression that evaluates to a [boolean][]:

<?code-excerpt "misc/lib/language_tour/control_flow.dart (if-else)"?>
```dart
if (isRaining()) {
  you.bringRainCoat();
} else if (isSnowing()) {
  you.wearJacket();
} else {
  car.putTopDown();
}
```

To learn how to use `if` in an expression context, see [Conditional expressions][]

### If-case

Dart `if` statements support `case` clauses followed by a [pattern][]: 

```dart
if (pair case [int x, int y]) return Point(x, y);
```

If the pattern (the list pattern `[int x, int y]` in the example above)
matches the value (`pair`), then the branch (`return Point(x, y)`)

> executes with any variables the pattern defines in scope (`x` and `y`).

Author: I was hoping someone would fix this! 😅

> Otherwise, the else branch is executed, if there is one:

Reviewer: **issue:** Write in active voice per [GSG](https://developers.google.com/style/voice).

```suggestion
Otherwise, Dart executes the `else` branch when possible.
```

```dart 
if (pair case [int x, int y]) {
  print('Was coordinate array $x,$y');
} else {
  throw FormatException('Invalid coordinates.');
}
```

If-case is a simple way to match and destructure against a
_single_ pattern. To test a value against _multiple_ patterns, use [switch](#switch).

## Switch

Switch statements evaluate a value expression against a series of cases.
Each case clause is a [pattern][] to match against the value being switched on,
followed by a series of statements in the case body to execute if the value

> matches that pattern. You can use [any kind of pattern][] for a case.

Author: I use backticks to refer to the `case` keyword, so if I'm talking about the code itself. But mostly when I'm talking about cases, it's just the concept, so no backticks. Other than that, done

Non-empty `case` clauses implicitly jump to the end of the switch after completion.
Other valid ways to end a non-empty `case` clause are a [`continue`][break],

> [`throw`][], or [`return`][] statement.

Author: Same about backticks for switch.

I'm a little torn on "implicitly"; I included it to kind of pose a counter to the previous behavior, which was not implicit (required `break` to end).  So it's providing an explanation to people who are used to break ("now it implicitly ends"), without being too on the nose for those who are new to Dart (like including a note about how it "used to be"). Lmk what you think.

> [`throw`][], or [`return`][] statement.

Reviewer: **comment:** Be direct. You can say:
```markdown
on-empty `case` clauses jump to the end of the `switch` after completion without requiring a `break`.
```

> [`throw`][], or [`return`][] statement.

Reviewer: **nit:** You mention the backticks but don't write what you mean. If you don't want to add them in this pass, don't. Let's land this.

> [`throw`][], or [`return`][] statement.

Author: Got it, using the more direct suggestion. 

I thought I wrote about the backticks in another comment, but I guess not. I'll keep them for now.

Use a `default` or wildcard `_` clause to execute code when no `case` clause matches:

```dart
var command = 'OPEN';
switch (command) {
  case 'CLOSED':
    executeClosed();
  case 'PENDING':
    executePending();
  case 'APPROVED':
    executeApproved();
  case 'DENIED':
    executeDenied();
  case 'OPEN':
    executeOpen();
  default:
    executeUnknown();
}
```

Empty cases fall through to the next case. For non-sequential fall-through,
you can use a [`continue` statement][break] and a label:

```dart
switch (command) {
  case 'OPEN':
    executeOpen();
    continue newCase;     // Continues executing at the newCase label.
  
  case 'DENIED':          // Empty case falls through.
  case 'CLOSED':
    executeClosed();      // Runs for both DENIED and CLOSED,
  
  newCase:
  case 'PENDING':
    executeNowClosed();   // Runs for both OPEN and PENDING.
}
```
For an empty case that does not fall through, use [`break`][break] for its body.

### Switch expressions

Switch expressions are similar to switch statements, but you can use them anywhere
you can use an expression. They produce a value based on the expression body of

> whichever case matches.

Author: Not a fan of breaking the sentence like that, removing the counter "but" waters down the difference. I'd argue that the sentence is not egregiously long, and two clause sentences are reasonable.

- Each case must have a body; there is no implicit fallthrough for empty cases.
- Case patterns are separated from their bodies using `=>` instead of `:`.
- Cases are separated by `,` (and an optional trailing `,` is allowed).
- Default cases can _only_ use `_`, instead of allowing both `default` and `_`.

Switch expressions allow you to rewrite a _statement_ like this:

```dart
// Where slash, star, comma, semicolon, etc., are constant variables...

switch (charCode) {
  case slash || star || plus || minus:    // Logical-or pattern
    token = operator(charCode);
  case comma || semicolon:                // Logical-or pattern
    token = punctuation(charCode);
  case >= digit0 && <= digit9:            // Relational and logical-and patterns
    token = number();
  default:
    throw invalid();
}
```

Into an _expression_, like this:

```dart
token = switch (charCode) {
  slash || star || plus || minus => operator(charCode),
  comma || semicolon => punctuation(charCode),
  >= digit0 && <= digit9 => number(),
  _ => throw invalid()
};
```

### Exhaustiveness checking

Exhaustiveness checking is a feature of switches that reports a compile-time

> error if it's possible for a value entering the switch to not match any of the cases.

Author: this is one of those things I'm hesitant to change now that we're past technical reviews. Exhaustiveness checking happens _to_ switches... I know that saying it's `switch` that's responsible for exhaustiveness checking would not go over well; it would probably be the compiler or something, but it's weird to bring up the compiler in the language tour... I changed it, when you see the next commit see what you think

```dart
bool? b = false;
// Non-exhaustive switch on bool?, missing case to match null possiblity:
switch (b) {
  case true: print('yes');
  case false: print('no');
}
```

A default case (`default` or `_`) covers all possible values, making a switch on
any type exhaustive.

[Enums][enum] and [sealed types][sealed] are particularly useful for switches
because, even without a default case, their possible values are known and fully
enumerable. Use the [`sealed` modifier][sealed] on a class to enable
exhaustiveness checking when switching over subtypes of that class:

```dart
sealed class Shape {
  double calculateArea();
}

class Square extends Shape { ... }
class Circle extends Shape { ... }

// ...

double calculateArea(Shape shape) =>
  switch (shape) {
    Square(length: var l) => l * l,
    Circle(radius: var r) => math.pi * r * r
  };
```

> If a new subclass of `Shape` is ever added, exhaustiveness 
> checking will tell you about the missing subtype, because your switch is incomplete.
> This allows you to use Dart in a somewhat 
> [functional algebraic datatype style](https://en.wikipedia.org/wiki/Algebraic_data_type). 

Reviewer: **issue:** Write in active voice per [GSG](https://developers.google.com/style/voice). 

```suggestion
If you add a new subclass of `Shape`,
exhaustiveness checking reports the missing subtype.
The check flags your `switch` as incomplete.
This allows you to use Dart in a somewhat 
[functional algebraic datatype style](https://en.wikipedia.org/wiki/Algebraic_data_type). 
```

Author: The benefit here is that if a subclass is added _unbeknownst to you_, exhaustiveness checking would alert you to it, so using "you" as the actor doesn't work for the selling point. I tried this:

> If anyone were to add a new subclass of `Shape`, this `switch` expression would 
be incomplete. Exhaustiveness checking would inform you of the missing subtype.

## Guard clause

<a id="when"></a>

> An optional guard clause can appear after any case clause
> (`if case`, and both `switch` statements and expressions)
> using the keyword `when`. Guards evaluate an arbitrary boolean expression _after_
> matching, allowing you to further constrain whether a case body should execute
> despite the case pattern matching the value.

Reviewer: **issues:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading). Avoid parentheticals. Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure).
```suggestion
To set an optional guard clause after a `case` clause, use the keyword `when`.
You can use this keyword with `if case`, and both `switch` statements and expressions.
Guards evaluate an arbitrary boolean expression _after_ matching.
This allows you to determine if a `case` body should execute
despite the `case` pattern matching the value.
```

### Case patterns

The pattern following a case, or case pattern, allows control flow to
either match and destructure the object being switched on, or continue execution
if the object doesn't match. Case patterns can be any kind of [pattern][]:

- **Constant patterns**: the object's value has to be equivalent to match.
  ```dart
  switch (number) {
    case 1:           // Matches if number == 1
      print("one");     
  }
  ```
- **List**, **map**, or **record patterns**: the elements are destructured if matched.
  ```dart
  switch (obj) {
    case [a, b]:                  // Matches if obj is a list with two elements
      print("$a and $b match");    
  }
  ```  
- **Variable patterns**: the matched value binds to a new variable.
  ```dart
  switch ((1, 2)) {
    case (var a, var b): ...     // Matches if 1 and 2 can be assigned to a and b
  }
  ```  
- **Logical** or **relational patterns**: the object matches if the comparison to
a constant returns `true`.
  ```dart
  switch (code) {
    case >= first && <= last: ...    
    // Matches if the value of code is between the constant values of first and last
  }
  ```  

> - **Null check** or **assert patterns**: the match is first on whether the object is not null, then on the value.

Reviewer 2: I would split null-check and null-assert patterns into two separate sections since they behave pretty differently on null. (Also, the latter are called "null assert" patterns not just "assert" patterns since they're specific to handling null.)

Also, the wording is pretty strange. Maybe: If the object is `null`, then doesn't match. Otherwise, matches the object against the inner pattern.

Author: Moved this section to the types reference page where we addressed this, resolving!

  ```dart
  switch (maybe) {
    case var s?:                 
    // Matches if maybe is not null, and then against the inner pattern "var s"
  }
  ```  

> - **Object patterns**: the match exposes getters on the object type.

Reviewer 2: Maybe "Matches if the object has the given named type. If so, then calls getters on the object and matches them against corresponding inner patterns.

Author: Also moved and addressed this one, resolving!

---
title: Branches 
description: Learn how to use branches to control the flow of your Dart code.
---

You can control the flow of your Dart code using branches:

-   `if`, `case`, and `else`
-   `switch` and `case` statements

> -   `switch` and `case` expressions

Reviewer 2: `else` is always contained and subordinate to an `if` statement/element. Likewise, `case` is always part of some surrounding `switch` statement. I wouldn't refer to them as statements or expressions on their own since they really aren't. Maybe:

```
You can control the flow of your Dart code using branching:

-  `if` statements and elements
-  `if-case` statements and elements
-   `switch` statements and expressions
```

And then just describe `else` and `case` as part of their respective contained things.

Other methods of manipulating control flow in Dart are:

-   [`for` loops][]
-   [`while` and `do while` loops][]
-   [`break` and `continue`][break]

> -   [`assert`][]

Reviewer 2: I'd remove `assert()` from here.

You can also affect control flow using [Exceptions][], like `try-catch` and `throw`.

> ## If else

Reviewer 2: I'd either leave off `else` here or add it to `## If case` below since both of them can either have or omit the `else` clause. Or maybe do:

```
## If and If-else

...

## If-case and If-case-else
```

> Dart supports `if` statements with optional `else` statements:

Reviewer 2: "`else` statements" -> "`else` clauses".

<?code-excerpt "misc/lib/language_tour/control_flow.dart (if-else)"?>
```dart
if (isRaining()) {
  you.bringRainCoat();
} else if (isSnowing()) {
  you.wearJacket();
} else {
  car.putTopDown();
}
```

> The statement conditions must be expressions

Reviewer 2: Maybe "The condition in parentheses after `if` must be an expression that evaluates to a [boolean][]."

that evaluate to [boolean values][boolean], nothing else.
To learn how to use `if` in an expression context, see [Conditional expressions][]

## If case

> Dart `if` statements support `case` clauses followed by a [case pattern][pattern]: 

Reviewer 3: I don't think there's a need to introduce a new concept of 'case patterns'. The patterns occurring in this syntactic context are just patterns.

Of course, we do have the notions of 'declaration context', 'matching context', 'assignment context', and you could argue that we have different kinds of patterns in those contexts. (For instance, an identifier pattern has a very different meaning in each of those contexts.) But it's quite reasonable to say that we just have patterns, and they can occur in each of those different contexts. With that, we don't need to introduce different words to denote patterns when they occur in different contexts. I think that's going to be more comprehensible.

Author: Great point, I'll incorporate this feedback throughout the page

```dart
if (pair case [int x, int y]) return Point(x, y);
```

If the case pattern (the list pattern `[int x, int y]` in the example above)
matches the case expression (`pair`), then the branch (`return Point(x, y)`)
executes with any variables the pattern defines in scope (`x` and `y`).

Otherwise, the else branch is executed, if there is one:

```dart 
if (pair case [int x, int y]) {
  print('Was coordinate array $x,$y');
} else {
  throw FormatException('Invalid coordinates.');
}
```

If-case is a simple way to match and destructure against a
_single_ pattern. To test a value against _multiple_ patterns, use [switch](#switch-case).

> ## Switch case

Reviewer 2: Just "## Switch".

Cases are the things inside the switch, but I think most people refer to the entire statement as a "switch" and then use "switch case" to refer to a single case within a switch. So, given:

```dart
switch (n) {
  case 1: ...
  case 2: ...
  default: ...
}
```

I'd say, "This switch statement matches the value of `n`. The first switch case checks if it's 1 and the second switch case check if it's 2. The default case handles any other numbers."

Switch cases (both [statements](#switch-statements) and [expressions](#switch-expressions))

> in Dart compare an object against one or more [case patterns][pattern]. 

Reviewer 2: Should this link work? It doesn't seem to.

Author: There's no patterns page yet, so it just goes to the home page for now. All of these dart 3 prs will be merged into the `v3.0` branch before `main`, so I was thinking I'd fill in the actual links in another pr on that branch once they're all in.

Case patterns allow control flow to

> either match and destructure the object being switched, or continue execution

Reviewer 3: Perhaps?:
```suggestion
either match and destructure the object being switched on, or continue execution
```
'Switching on' an object is a rather common phrase. Some people would say that object is the 'scrutinee', but perhaps that's too specialized (it's _the_ thing to say in the functional programming language community).

Author: Agreed it's a little too specialized here, but I'll definitely be using "scrutinee" in the functional-style-dart blog post I'm writing!

> if the object doesn't match. They can be any kind of [refutable pattern][]:

Reviewer 3: They can be refutable as well as irrefutable patterns. So it isn't going to help anyone to suggest that they can't be irrefutable. However, saying that it's 'any kind of pattern' and then 'including a [refutable pattern][]' might be helpful.

Author: Aha, I took the point of _"you can't use refutable patterns in places that don't do control flow"_ in the spec to mean the inverse (_you can't use irrefutable patterns in places that do control flow_) as well. But, reading again, the spec doesn't say that anywhere. Thanks for catching!

> if the object doesn't match. They can be any kind of [refutable pattern][]:

Reviewer 2: "Continue execution if the object doesn't match" is a little unclear about *where* it continues to. I'd probably merge the "Switch case" and "Switch statement" sections together and say something like:

> A switch statement allows you to test a single value against a series of patterns and execute code for the first pattern that matches. A switch statement has evaluates a *matched value expression*. Then it has a series of *cases*. Each case has a pattern, and a block of code to execute if the pattern matches. After evaluating the matched value expression, each pattern is matched against the value. The first one that matches has its block of code executed. After that, execution jumps to the end of the switch statement.

Note also that switch case patterns can be irrefutable patterns too. There are some places in the language that disallow refutable patterns, but irrefutable patterns are always allowed anywhere a pattern can appear.

Author: I was thinking about merging those sections together myself but wasn't sure if it was important to distinguish "statements"; thanks for the confirmation!

  ```dart
  switch (obj) {
    case [a, b]:                  // Matches if obj is a list with two elements
      print("$a and $b match");    
  }
  ```  
- **Variable patterns**: the matched value binds to a new variable.
  ```dart
  switch ((1, 2)) {
    case (var a, var b): ...     // Matches if 1 and 2 can be assigned to a and b
  }
  ```  
- **Logical** or **relational patterns**: the object matches if the comparison to a constant returns `true`.
  ```dart
  switch (code) {
    case >= first && <= last: ...    
    // Matches if code's value is between the constant values of first and last
  }
  ```  
- **Null check** or **assert patterns**: the match is first on whether the object is not null, then on the value.
  ```dart
  switch (maybeString) {
    case var s?:                 
    // Matches if maybe is not null, and then against the inner pattern "var s"
  }
  ```  
- **Object patterns**: the match exposes getters on the object type.
  ```dart
  switch (Shape shape) {
    case Rect(width: var w, height: var h): ...
    // Matches if shape is of type Rect, and then against the properties of Rect
  }
  ```  

> Values destructured by a case pattern become local variables, only available in the
> scope of that clause. 

Reviewer 3: People might not have an idea about how that scope is determined. Perhaps:
```suggestion
Values destructured by a pattern in a case become local variables, only available in the
body of that case (or group of cases, when one body is shared by several cases). 
```

Author: I'm going to omit the parenthetical because I think "the body of a case", regardless of how that body ended up associated with that case, reasonably covers shared case bodies

> scope of that clause. 

Reviewer 2: Probably "clause" -> "case body".

Other valid ways to end a non-empty `case` clause are a [`continue`][break],
[`throw`][], or [`return`][] statement.

Use a `default` or wildcard `_` clause to execute code when no `case` clause matches:

```dart
var command = 'OPEN';
switch (command) {
  case 'CLOSED':
    executeClosed();
  case 'PENDING':
    executePending();
  case 'APPROVED':
    executeApproved();
  case 'DENIED':
    executeDenied();
  case 'OPEN':
    executeOpen();
  default:
    executeUnknown();
}
```

Empty cases fall through to the next case. For an empty case that does not fall
through, use [`break`][break] for its body:

```dart
var command = 'CLOSED';
switch (command) {
  case 'CLOSED': // Empty case falls through.
  case 'NOW_CLOSED':
    // Runs for both CLOSED and NOW_CLOSED.
    executeNowClosed();
}
```

For non-sequential fall-through, you can use a [`continue` statement][break] and
a label:

```dart
var command = 'CLOSED';
switch (command) {
  case 'CLOSED':
    executeClosed();
    continue nowClosed;
  // Continues executing at the nowClosed label.

  nowClosed:
  case 'NOW_CLOSED':
    // Runs for both CLOSED and NOW_CLOSED.
    executeNowClosed();
}

> ```

Reviewer 2: Continue to label is so rarely used, I would be tempted to omit it from the guide completely. (Honestly, I was hoping to remove it from the language entirely in Dart 3.0, but, alas, I wasn't able to pull it off.)

Reviewer 4: i would prefer to document the entire language than have hidden features. People will come across code that uses this, and then go looking for what it means, and the language tour is the only approachable language reference we provide, so if a feature is missing from the guide they'll be lost.

Author: Another case for needing fuller reference type docs per language topic, tracking here: #4783 

I'll leave the example for now, but will add this to the TODO list for reference docs.

### Switch expressions

_Switch expressions_ write a result value determined by a switch

> directly into a variable or return statement. 

Reviewer 2: You can use a switch expression anywhere an expression can appear so "directly into a variable or return statement" is probably too specific. For example, this is fine:

```dart
print(switch (n) { 1 => 'one', 2 => 'two', _ => 'other'});
```

How about:

> A switch expression is similar to a switch statement, but can be used where an expression is expected and produces a value based on the expression body of the case that matched.

The syntax rules for a switch expression are:
- The expression body is a series of case patterns.
- Cases _do not_ use the `case` keyword.
- Case patterns are followed by `=>`, and then a case body. 
- Case bodies terminate with a comma `,`.

> - The switch expression body terminates with a semicolon `;`.

Reviewer 3: That's not a property of the switch expression or its body, it is a requirement in some situations of the enclosing construct: A statement consisting of an expression ends in a semicolon (e.g., `cond ? 1 : switch (x) {...};` needs the semicolon because it's a statement of the form `e;` where `e` is an expression). In other situations the `}` could be followed by a `,`, say, if the switch is an actual argument in a function invocation.

So, maybe:
```suggestion
- The switch expression body may be followed by a semicolon `;`, e.g., if it is part of a declaration or a statement.
```

Author: Removed the line entirely in favor of just saying it can be used anywhere an expression can be

> - The switch expression body terminates with a semicolon `;`.

Reviewer 2: You may see this in some examples, but that's not part of the switch expression, it's a result of the context where the expression appears. In:

```dart
var x = switch (n) { 1 => 'one', _ => 'other'};
```

The trailing `;` is part of the variable declaration statement, not the switch, the same as in:

```dart
var x = 3;
```

There's no semicolon at the end in, for example:

```dart
print(switch (n) { 1 => 'one', _ => 'other' })
```

> - The wildcard pattern `_` catches remaining values instead of `default`.

Reviewer 2: You can actually use `_` as a default case in switch statements too. (Whether that will become idiomatic or not is anybody's guess.)

Maybe something like:

> Switch expression syntax is slightly different from switch statements:
>
> - Cases *don't* start with the `case` keyword.
> - The case pattern is separated from its body using `=>` instead of `:`.
> - The case body is a single expression instead of a series of statements. Every case must have a body -- there is no implicit fallthrough for empty cases.
> - Cases are separated by `,` (and an optional trailing `,` is allowed).

This allows you to rewrite a 

> _statement_ like this:

Reviewer 3: ```suggestion
_statement_ like this (where `slash`, `star`, ... are constant variables):
```
I think it's somewhat hard to understand what it is doing if we don't know that they must be constant variables.

Author: I can add their declarations to the code block.

```dart
switch (charCode) {
  case slash when nextCharCode == slash:   // Constant pattern with guard
    skipComment();
  case slash || star || plus || minus:     // Logical-or pattern
    operator(charCode);
  case >= digit0 && <= digit9:             // Relational and logical-and patterns
    number();
  default:
    invalid();
}
```

Into an _expression_, like this:

```dart
var token = switch (charCode) {
  slash when nextCharCode == slash => skipComment(),
  slash || star || plus || minus   => operator(charCode),
  >= digit0 && <= digit9           => number(),

>   _                                => throw invalid()

Reviewer 3: This looks tempting, but Dart standard formatting _never_ makes an attempt to use alignment such that specific tokens (`=>`) end up looking like a vertical line.

Perhaps we shouldn't give anybody the impression that they should align their code like that?

Reviewer 4: or maybe we should fix the formatter to do it :-)
this is way more readable than not doing it...

Author: I'll change it back to normal but I wish this formatting was ok to use in the docs! It's so much clearer

};

> ```

Reviewer 2: These examples aren't quite 1-1. How about:

---

This allows you to rewrite a 
_statement_ like this:

```dart
switch (charCode) {
  case slash || star || plus || minus: // Logical-or pattern
    token = operator(charCode);
  case comma || semicolon: // Logical-or pattern
    token = punctuation(charCode);
  case >= digit0 && <= digit9: // Relational and logical-and patterns
    token = number();
  default:
    throw invalid();
}
```

Into an _expression_, like this:

```dart
token = switch (charCode) {
  slash || star || plus || minus => operator(charCode),
  comma || semicolon => punctuation(charCode),
  >= digit0 && <= digit9 => number(),
  _ => throw invalid()
};
```

### Exhaustiveness checking

> Exhaustiveness checking is a feature of switches that raises a compile-time error

Reviewer 2: I try to only use "raise" to refer to throwing runtime exceptions. (For comparison, the keyword to throw an exception in Python and Ruby is `raise`.) Maybe "that reports a compile-time error"?

> if a switch is not exhaustively covering all possible values that could flow through it.

Reviewer 2: Defining exhaustiveness in terms of "exhaustively" may not be very clear. :)

Maybe "reports a compile-time error if it's possible for a value entering the switch to not match any of the case patterns".

Author: I went with your version, except removed "patterns" from the end (just "cases") because of [Erik's idea](https://github.com/dart-lang/site-www/pull/4780#discussion_r1176761413) to not get into "case patterns" terminology too much here

The compiler can only know if a set of cases is
exhaustive if it is soundly aware of all possible enumerations of that type. So,

> exhaustiveness checking works only when switching on types that are fully enumerable:

Reviewer 3: ```suggestion
exhaustiveness checking on switch _statements_ works only when switching on types that are fully enumerable:
```

Author: removed this section per [bob's comment](https://github.com/dart-lang/site-www/pull/4780#discussion_r1176833411)

> exhaustiveness checking works only when switching on types that are fully enumerable:

Reviewer 2: Exhaustiveness does actually work for all kinds of types. (Otherwise switch expressions would be much less useful because they *have* to be exhaustive. Otherwise, you might not know what value the expression evaluates to.)

The reason that enums and sealed types are particularly useful for switches is that they allow you to "partition" the matched value in a way such that multiple cases each of which handle only a subset of the possible values can provably cover all possible values.

For other types, to be exhaustive, you basically need to have a single case that covers everything. So in:

```dart
int n = ...
var name = switch (n) {
  1 => 'one',
  2 => 'two',
  _ => 'other'
};
```

The switch is exhaustive because the final `_` pattern covers all possible ints.

I'd probably just remove this section. I don't think we go into a lot of detail about how type inference and flow analysis work and those are similar where there's a decent amount of algorithm sophistication going on under the hood and we just expect that it aligns with user intuition.

So if they have an intuition about the set of all possible values going into a switch and a good intuition about which values the set of patterns covers then exhaustiveness checking should mostly align with that. The actual mechanics of it get pretty complex.

- [Boolean][boolean] types, because there are only two possible values, `true` and `false`.
- [Enum][enum] types, because they are always a fixed number of constant values.

> - [Sealed][sealed] types, because they are classes with a closed hierarcy of subtypes in the same library.

Reviewer 2: "hierarchy".

```dart
bool? b = false;
// Non-exhaustive switch on bool?, missing case to match null possiblity:
switch (b) {
  case true: print('yes');
  case false: print('no');
}
```

Use the [`sealed` modifier][sealed] on a class to enable exhaustiveness checking when 
switching over subtypes of that class:

```dart
sealed class Shape {
  double calculateArea();
}

class Square extends Shape { ... }
class Circle extends Shape { ... }

// ...

double calculateArea(Shape shape) =>
  switch (shape) {
    Square(length: var l) => l * l,
    Circle(radius: var r) => math.pi * r * r
  };
```

If a new subclass of `Shape` is ever added, exhaustiveness 

> checking will warn you about the missing subtype, because your switch is incomplete.

Reviewer 2: "warn" -> "tell". (Since it's an error, not a warning.)

This allows you to use Dart in a somewhat 

> [functional algrbraic datatype style](https://en.wikipedia.org/wiki/Algebraic_data_type). 

Reviewer 2: "algebraic".

> 

Reviewer 3: Then we'd need to say that switch expressions are always required to be exhaustive, and that this is possible even with types that aren't enumerable if there is a case that covers all possible values (such as a default case or a case like `var x` which will match every value of the scrutinee and give `x` the same declared type as the static type of the scrutinee expression).

> ## When

Reviewer 2: I'd name the heading "Guard clause" since that's what it is.

Reviewer 4: maybe "Guard clause (`when`)"

### Switch expressions

> A `switch` expression is similar to a `switch` statement, but you can use them

Reviewer: **suggestion:** Active voice.
```suggestion
A `switch` expression works like a `switch` statement, but you can use them
```

Author: Gonna skip, the difference is more in how they work. Their similarities are also in how they work, but I just don't want to start with that.

### Exhaustiveness checking

> Exhaustiveness checking is a feature that reports a compile-time

Reviewer: **nit:** Consider shortening this.
```suggestion
Exhaustiveness checking reports a compile-time
```

Author: Hmm.. I'm gonna keep it because "exhaustiveness checking" sounds like a verb, but it's kind of actually a "thing"... I guess. I think clarity helps here.
